chasm/lib/activity/handler.go

```
@@ -51,3 +53,177 @@ func (h *handler) StartActivityExecution(ctx context.Context, req *activitypb.St
+			lastSeenRef.BusinessID != req.GetFrontendRequest().GetActivityId() ||
+			lastSeenRef.EntityID != req.GetFrontendRequest().GetRunId() {
+			// token is inconsistent with request
+			return nil, nil, serviceerror.NewInvalidArgument("invalid long poll token")
+		}
+	} else {
+		// This ref will compare less than currentRef in the comparison below.
+		lastSeenRef = chasm.NewComponentRef[*Activity](chasm.EntityKey{
+			NamespaceID: req.GetNamespaceId(),
+			BusinessID:  req.GetFrontendRequest().GetActivityId(),
+			EntityID:    req.GetFrontendRequest().GetRunId(),
+		})
+	}
+
+	// PollComponent will return an error if lastSeenRef is not consistent with the entity
+	// transition history on this shard, or if the state on this shard is behind the ref after a
+	// reload.
+	// TODO(dan): retryability of these errors
+	return chasm.PollComponent(
+		ctx,
+		lastSeenRef,
+		func(
+			a *Activity,
+			ctx chasm.Context,
+			req *activitypb.PollActivityExecutionRequest,
+		) (*activitypb.PollActivityExecutionResponse, bool, error) {
+			// TODO(dan): we're walking the tree to construct a ref when all we want here is the
+			// root/entity VT. Would it make sense for Context to provide access to root node?
+			currentRefBytes, err := ctx.Ref(a)
+			if err != nil {
+				return nil, false, err
+			}
+			currentRef, err := chasm.DeserializeComponentRef(currentRefBytes)
+			if err != nil {
+				return nil, false, err
+			}
+
+			if lastSeenRef.EntityID != currentRef.EntityID {
+				// The runID from the token doesn't match this shard's state. We return immediately,
+				// on the basis that this constitutes a state change. If the runID from the token is
+				// ahead of this shard's state then this will be detected by shard ownership or
+				// staleness checks and the caller will receive an error.
+				response, err := a.buildPollActivityExecutionResponse(ctx, req)
+				if err != nil {
+					return nil, true, err
+				}
+				return response, true, nil
```

maybe I'm misunderstanding, but if the runIDs don't match, should we return an error here?

---

ok, that behavior makes sense